libs/compute_api/src/spec.rs

```diff
@@ -3,36 +3,78 @@
 //! The spec.json file is used to pass information to 'compute_ctl'. It contains
 //! all the information needed to start up the right version of PostgreSQL,
 //! and connect it to the storage nodes.
-use serde::Deserialize;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
 use std::collections::HashMap;
+use utils::id::{TenantId, TimelineId};
+use utils::lsn::Lsn;
 
 /// String type alias representing Postgres identifier and
 /// intended to be used for DB / role names.
 pub type PgIdent = String;
 
 /// Cluster spec or configuration represented as an optional number of
 /// delta operations + final cluster state description.
-#[derive(Clone, Deserialize)]
+#[serde_as]
+#[derive(Clone, Deserialize, Serialize)]
 pub struct ComputeSpec {
     pub format_version: f32,
     pub timestamp: String,
+    /// ID of the operation that launched this compute. Not used by 'compute_ctl',
+    /// just for debugging purposes.
     pub operation_uuid: Option<String>,
     /// Expected cluster state at the end of transition process.
     pub cluster: Cluster,
     pub delta_operations: Option<Vec<DeltaOp>>,
 
+    // Information needed to connect to the storage layer.
+    //
+    // `tenant_id`, `timeline_id` and `pageserver_connstring` are always needed.
+    //
+    // If Lsn == None, this is a primary endpoint that continues writing WAL at
+    // the end of the timeline. If 'lsn' is set, this is a read-only node
+    // "anchored" at that LSN. 'safekeeper_connstrings' must be non-empty for a
+    // primary.
+    //
+    // For backwards compatibility, the control plane may leave out all of
+    // these, and instead set the "neon.tenant_id", "neon.timeline_id",
+    // etc. GUCs in cluster.settings. TODO: This is deprecated; once the control
+    // plane has been updated to fill these fields, we can make these non
+    // optional.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub tenant_id: Option<TenantId>,
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub timeline_id: Option<TimelineId>,
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    pub lsn: Option<Lsn>,
+    pub pageserver_connstring: Option<String>,
+    pub safekeeper_connstrings: Vec<String>,
+
+    /// If set, 'storage_auth_token' is used as the password to authenticate to
+    /// the pageserver and safekeepers.
     pub storage_auth_token: Option<String>,
 
+    /// W3C trace context of the launch operation, for OpenTelemetry tracing
     pub startup_tracing_context: Option<HashMap<String, String>>,
 }
 
-#[derive(Clone, Deserialize)]
+#[derive(Clone, Deserialize, Serialize)]
 pub struct Cluster {
     pub cluster_id: String,
     pub name: String,
     pub state: Option<String>,
     pub roles: Vec<Role>,
     pub databases: Vec<Database>,
+
+    /// Desired contents of 'postgresql.conf' file. (The 'compute_ctl'
+    /// tool may add additional settings to the final file.)
+    pub postgresql_conf: Option<String>,
```

Comment:
> IMHO it would be nice to preserve comments and other formatting in postgresql.conf. That's why I would like to keep it as a flat text string. I know we don't currently do that in the control plane, and we don't really give users a way to see postgresql.conf as a file, but that might change in the future too.

When we allow changing settings, I'd imagine it to be a nice UI (we actually had it in some very old UI version) with separate input per settings, selectors for enums, validations, hints what each setting mean and so on, not just bulk text editor. It looks way too complicated to send the whole `postgresql.conf` text to the API server, deserialize it, check that some value exceeds the limit or it's not allowed to change some setting and send error back just to render a nice error.

I don't have that strong opinion, just don't really like that caller (control-plane / console / someone else) has to know `postgresql.conf` escaping rules: what should be wrapped with commas, which commas to use, what is the escape sequence. The lower this logic is buried the better

Reply:
> When we allow changing settings, I'd imagine it to be a nice UI (we actually had it in some very old UI version) with separate input per settings, selectors for enums, validations, hints what each setting mean and so on, not just bulk text editor.

Makes sense.

> It looks way too complicated to send the whole `postgresql.conf` text to the API server, deserialize it, check that some value exceeds the limit or it's not allowed to change some setting and send error back just to render a nice error.

Yeah, that would be tedious. I'd imagine that we keep the settings in the current key-value format in the control plane, and construct the postgresql.conf file at the last moment, when we launch the compute node. That might be an alternative or an addition to storing the raw postgresql.conf blob.

We might also want to support ALTER SYSTEM in PostgreSQL, in a way that survives compute restarts...

But this seems unrelated to where convert the key-value pairs to a postgresql.conf file.

> I don't have that strong opinion, just don't really like that caller (control-plane / console / someone else) has to know `postgresql.conf` escaping rules: what should be wrapped with commas, which commas to use, what is the escape sequence. The lower this logic is buried the better

The postgresql.conf escaping rules are pretty simple, really:
- Every line is like `name='value'`
- If there is a single quote in the value, duplicate it
- If there is a backslash in the value, duplicate it

That's it.

Now, if you want to do further validation, like which fields take a unit, which enum values are accepted, what are the accepted ranges and combinations for the settings etc, that's a lot more complicated. But if we want to catch those errors before launching PostgreSQL, it's too late to verify those things when we're starting up the compute.

> After sleeping with it I realized that to make it working we will need to pass each setting type as well. Similar to what we do now.

That's not really necessary. If a GUC is of an integer type, you can write it as `key=123`, with no quotes. But you _can_ use quotes like you do with strings, i.e. `key='123'` works too. You don't need to know the types when you construct the postgresql.conf contents.

As a side note, when we present the settings to users, we should definitely support units. Being able to write `work_mem='10 MB'` is massively more user-friendly than `work_mem='10485760'`. We could convert the former to the latter when we build the postgresql.conf file, but we could also let PostgreSQL handle it. If we let PostgreSQL handle it, we should treat '10 MB' the same as string settings.

> And if for cloud control-plane we already have it, demanding the same from neon tests / neon_local is meh, I agree. Let's keep both ways of passing settings for some time, though

Ok, I can live with that.